2025/06/bash-expansion-examples.md

```diff
@@ -0,0 +1,114 @@
+featured:
+  image_url:
+github_issue_number: 
+tags:
+- linux
+- shell
+- tips
+---
+
+For any project, you need a quick and efficient way to wrangle your files. As a Unix user, Bash and Zsh provide are powerful tools to help achieve this.
+
+I recently needed to rename a file so that all its underscores were replaced with dash characters, to match the convention of the project. I could do this manually pretty quickly, but I knew there was a bash built-in one-liner waiting to be discovered, so I went down the rabbit hole to learn about Bash's [shell expansions](https://www.gnu.org/software/bash/manual/html_node/Shell-Expansions.html) and [history expansion](https://www.gnu.org/software/bash/manual/html_node/History-Interaction.html). See the "history expansion" section for how I solved the underscore/dash issue.
+
+Bash has seven types of expansion:
+
+- brace expansion
+- tilde expansion
+- parameter and variable expansion
+- command substitution
+- arithmetic expansion
+- word splitting
+- filename expansion 
+
+The documentation is good and concise for each of these, so rather than try to recreate it, I'll go over examples of how I use some of them.
+
+### Shell parameter expansion
+
+#### Example: batch converting images to WebP
+
+I use parameter expansion frequently while maintaining this blog. We serve images in [WebP format](/blog/2014/01/webp-images-experiment-on-end-point/), so I generally loop over all the JPEGs and/or PNGs (after cropping and/or scaling) and convert them using [cwebp](https://developers.google.com/speed/webp/docs/cwebp):
+
+```bash
+for f in *.png; do
+  cwebp -q 80 $f -o ${f%png}webp
+done
+```
+
+This makes use of parameter expansion (`${}`) and the `%` character, which deletes the shortest occurrence of the *word* it precedes (`png`) from the end of the *parameter* (`$f`, the filename). Then, you can insert a string immediately after the expansion and Bash will concatenate them together, making a new output filename.
+
+If you use `%%` instead, it deletes the longest occurrence of the *word* from the end, instead of the shortest.
+
+You can use the `#` character to remove a *word* from the beginning of the string, sort of the converse of `%` removes a match from the end.
+
+> The formal definition of *word* in [the Bash manual](https://www.gnu.org/software///bash/manual/bash.html) is "A sequence of characters treated as a unit by the shell. Words may not include unquoted metacharacters."
+>
+> I looked this up so I would know the distinction between *words* and [*parameters*](https://www.gnu.org/software///bash/manual/bash.html#Shell-Parameters). Parameters are more specific: mainly just variables, positional parameters, and a few special parameters.
+
+#### Example: counting file endings
+
+I haven't used this one as often, but you can use `##` to count the number of each file ending in the current directory:
+
+`${file/pattern/replacement}`
```

Comment:
It looks like this line might have been intended as an alternative for the section above? Or perhaps was for an alternate example? It doesn't seem to fit here.

FWIW I use that version much more often than the `%`, as it (at least to me) makes it more clear what's happening when doing a string replacement, a la `${f/png/webp}` or such. Also it doesn't do anything if "png" isn't found, whereas the above always appends "webp"... Which may be better anyway, depending on what you want to happen in that case. :)